// licia/Dispatcher.js
var Class = require('./Class');
var uniqId = require('./uniqId');

exports = Class({
    initialize: function Dispatcher() {
        this._callbacks = {};
        this._isDispatching = false;
        this._isHandled = {};
        this._isPending = {};
    },
    dispatch: function dispatch(payload) {
        this._startDispatching(payload);

        for (var id in this._callbacks) {
            if (this._isPending[id]) continue;

            this._invokeCb(id);
        }

        this._stopDispatching();
    },
    register: function register(cb) {
        var id = uniqId('ID_');
        this._callbacks[id] = cb;
        return id;
    },
    waitFor: function waitFor(ids) {
        for (var i = 0, len = ids.length; i < len; i++) {
            var id = ids[i];
            if (this._isPending[id]) continue;

            this._invokeCb(id);
        }
    },
    unregister: function unregister(id) {
        delete this._callbacks[id];
    },
    isDispatching: function isDispatching() {
        return this._isDispatching;
    },
    _startDispatching: function _startDispatching(payload) {
        for (var id in this._callbacks) {
            this._isPending[id] = false;
            this._isHandled[id] = false;
        }

        this._pendingPayload = payload;
        this._isDispatching = true;
    },
    _stopDispatching: function _stopDispatching() {
        delete this._pendingPayload;
        this._isDispatching = false;
    },
    _invokeCb: function _invokeCb(id) {
        this._isPending[id] = true;

        this._callbacks[id](this._pendingPayload);

        this._isHandled[id] = true;
    }
});

module.exports = exports;
