// licia/jsonp.js
var loadJs = require('./loadJs');
var defaults = require('./defaults');
var noop = require('./noop');
var uniqId = require('./uniqId');
var query = require('./query');

exports = (function(_exports) {
    function exports(_x) {
        return _exports.apply(this, arguments);
    }

    exports.toString = function() {
        return _exports.toString();
    };

    return exports;
})(function(opts) {
    defaults(opts, exports.settings);
    var name = opts.name || uniqId('jsonp'),
        param = opts.param,
        timeout = opts.timeout,
        error = opts.error,
        success = opts.success,
        complete = opts.complete,
        data = opts.data,
        url = opts.url,
        timer,
        isTimeout = false;

    if (timeout > 0) {
        timer = setTimeout(function() {
            isTimeout = true;
            error(new Error('Timeout'));
            complete();
        }, timeout);
    }

    window[name] = function(data) {
        success(data);
        complete();
        window[name] = noop;
    };

    data[param] = name;
    data = query.stringify(data);
    url += url.indexOf('?') > -1 ? '&' + data : '?' + data;
    loadJs(url, function(isLoaded) {
        if (isTimeout) return;
        if (timer) clearTimeout(timer);

        if (!isLoaded) {
            error(new Error());
            complete();
        }
    });
});

exports.settings = {
    data: {},
    param: 'callback',
    success: noop,
    error: noop,
    complete: noop,
    timeout: 0
};

module.exports = exports;
