// licia/PseudoMap.js
var Class = require('./Class');
var root = require('./root');
var detectMocha = require('./detectMocha');
var defineProp = require('./defineProp');
var keys = require('./keys');
var each = require('./each');
var isArr = require('./isArr');
var isUndef = require('./isUndef');

if (root.Map && !detectMocha()) {
    exports = root.Map;
} else {
    exports = Class({
        initialize: function PseudoMap(data) {
            this.clear();
            var self = this;
            defineProp(this, 'size', {
                get: function get() {
                    return keys(self._data).length;
                },
                set: function set() {},
                enumerable: true,
                configurable: true
            });

            if (data instanceof exports) {
                data.forEach(function(val, key) {
                    this.set(key, val);
                }, this);
            } else if (isArr(data)) {
                each(
                    data,
                    function(val) {
                        this.set(val[0], val[1]);
                    },
                    this
                );
            }
        },
        forEach: function forEach(fn, ctx) {
            each(
                this._data,
                function(val, key) {
                    fn.call(this, val, key);
                },
                ctx
            );
        },
        has: function has(key) {
            return !isUndef(this._data[key]);
        },
        get: function get(key) {
            return this._data[key];
        },
        set: function set(key, val) {
            this._data[key] = val;
        },
        delete: function _delete(key) {
            delete this._data[key];
        },
        clear: function clear() {
            this._data = {};
        }
    });
}

module.exports = exports;
