// licia/ReduceStore.js
var Class = require('./Class');
var clone = require('./clone');
var remove = require('./remove');

exports = Class({
    initialize: function ReduceStore(reducer, initialState) {
        this._reducer = reducer;
        this._state = initialState;
        this._curListeners = [];
        this._nextListeners = this._curListeners;
    },
    subscribe: function subscribe(listener) {
        var isSubscribed = true;

        this._ensureCanMutateNextListeners();

        this._nextListeners.push(listener);

        var self = this;
        return function() {
            if (!isSubscribed) return;
            isSubscribed = false;

            self._ensureCanMutateNextListeners();

            remove(self._nextListeners, function(val) {
                return val === listener;
            });
        };
    },
    dispatch: function dispatch(action) {
        this._state = this._reducer(this._state, action);
        var listeners = (this._curListeners = this._nextListeners);

        for (var i = 0, len = listeners.length; i < len; i++) {
            listeners[i]();
        }

        return action;
    },
    getState: function getState() {
        return this._state;
    },
    _ensureCanMutateNextListeners: function _ensureCanMutateNextListeners() {
        if (this._nextListeners === this._curListeners) {
            this._nextListeners = clone(this._curListeners);
        }
    }
});

module.exports = exports;
